library/init/category/lift.lean

```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2016 Gabriel Ebner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner, Sebastian Ullrich
+
+Classy functions for lifting monadic actions of different shapes.
+
+This theory is roughly modeled after the Haskell 'layers' package https://hackage.haskell.org/package/layers-0.1.
+Please see https://hackage.haskell.org/package/layers-0.1/docs/Documentation-Layers-Overview.html for an exhaustive discussion of the different approaches to lift functions.
+-/
+prelude
+import init.function init.coe
+import init.category.monad
+import init.util
+
+universes u v w
+
+/-- A function for lifting a computation from an inner monad to an outer monad.
+    Like [MonadTrans](https://hackage.haskell.org/package/transformers-0.5.5.0/docs/Control-Monad-Trans-Class.html),
+    but `n` does not have to be a monad transformer.
+    Alternatively, an implementation of [MonadLayer](https://hackage.haskell.org/package/layers-0.1/docs/Control-Monad-Layer.html#t:MonadLayer) without `layerInvmap` (so far). -/
+class has_monad_lift (m : Type u → Type v) (n : Type u → Type w) :=
+(monad_lift {} : ∀ {α}, m α → n α)
+
+/-- The reflexive-transitive closure of `has_monad_lift`.
+    `monad_lift` is used to transitively lift monadic computations such as `state_t.get` or `state_t.put s`.
+    Corresponds to [MonadLift](https://hackage.haskell.org/package/layers-0.1/docs/Control-Monad-Layer.html#t:MonadLift). -/
+class has_monad_lift_t (m : Type u → Type v) (n : Type u → Type w) :=
+(monad_lift {} : ∀ {α}, m α → n α)
+
+export has_monad_lift_t (monad_lift)
+
+/-- A coercion that may reduce the need for explicit lifting.
+    Because of [limitations of the current coercion resolution](https://github.com/leanprover/lean/issues/1402), this definition is not marked as a global instance and should be marked locally instead. -/
+@[reducible] def has_monad_lift_to_has_coe {m n} [has_monad_lift_t m n] {α} : has_coe (m α) (n α) :=
+⟨monad_lift⟩
+
+instance has_monad_lift_t_trans (m n o) [has_monad_lift n o] [has_monad_lift_t m n] : has_monad_lift_t m o :=
+⟨λ α ma, has_monad_lift.monad_lift (monad_lift ma : n α)⟩
+
+instance has_monad_lift_t_refl (m) : has_monad_lift_t m m :=
+⟨λ α, id⟩
+
+@[simp] lemma monad_lift_refl {m : Type u → Type v} {α} : (monad_lift : m α → m α) = id := rfl
+
+
+/-- A functor in the category of monads. Can be used to lift monad-transforming functions.
+    Based on pipes' [MFunctor](https://hackage.haskell.org/package/pipes-2.4.0/docs/Control-MFunctor.html),
+    but not restricted to monad transformers.
+    Alternatively, an implementation of [MonadTransFunctor](http://duairc.netsoc.ie/layers-docs/Control-Monad-Layer.html#t:MonadTransFunctor). -/
+class monad_functor (m m' : Type u → Type v) (n n' : Type u → Type w) :=
```

Comment:
AFAICT, it has a single application in the PR: `zoom`.
Do you have other (more convincing) applications for it?

Reply:
Ok, it is also used to implement `with_reader`.